### PR TITLE
Feature/29 audit trails

### PR DESCRIPTION
**Pull request by lokm01**
This PR implements the Audit Trails

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/rest/models/ChangedFieldsUpdateTransformResult.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  fieldName: String,
+  value1: T,
+  value2: T
+)
+
+case class ChangedFieldsUpdateTransformResult[C](
+  updatedEntity: C,
+  fields: Seq[ChangedField[Any]]
+) {
+    /** Generates list of fields, for which two values do not match - i.e have been changed
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/rest/models/ChangedFieldsUpdateTransformResult.scala]: Style -> align tabs

File: menas/src/main/scala/za/co/absa/enceladus/rest/services/VersionedModelService.scala [hunk 2]
```diff
@@ -52,28 +55,48 @@ abstract class VersionedModelService[C <: VersionedModel](versionedMongoReposito
...
-          getLatestVersion(itemName)
-        }
-      case None => Future.successful(None)
-    }
+  private[services] def update(username: String, itemName: String, itemVersion: Int, auditMessage: String)(transform: C => ChangedFieldsUpdateTransformResult[C]): Future[Option[C]] = {
+    for {
+      version <- getVersion(itemName, itemVersion)
+      transformed <- if(version.isEmpty) Future.failed(NotFoundException(s"Version $itemVersion of $itemName not found"))
+                      else Future.successful(transform(version.get))
+      update <- versionedMongoRepository.update(username, transformed.updatedEntity)
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/rest/services/VersionedModelService.scala]: Style - when you are aligning other `<-`, you could these too

Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/rest/services/VersionedModelService.scala]: yea... not currently supported by scalariform.. Not gonna maintain this manually

File: menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala [hunk 3]
```diff
@@ -21,19 +21,31 @@ import za.co.absa.enceladus.model.conformanceRule.ConformanceRule
...
-      latest
+    super.update(username, dataset.name, dataset.version, "Dataset Updated.") { latest =>
+      val updated = latest
         .setSchemaName(dataset.schemaName)
         .setSchemaVersion(dataset.schemaVersion)
         .setHDFSPath(dataset.hdfsPath)
         .setHDFSPublishPath(dataset.hdfsPublishPath)
         .setDescription(dataset.description).asInstanceOf[Dataset]
+
+      ChangedFieldsUpdateTransformResult(updatedEntity = updated, fields = Seq(
```
Comment (GeorgiChochov) [inline menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala]: How about having a method in each entity `getUpdateChanges(updated: VersionedModel): ChangedFieldsUpdateTransformResult` which produces this for you when you call `latest.getUpdateChanges(updated)`?

File: menas/src/main/scala/za/co/absa/enceladus/rest/models/ChangedFieldsUpdateTransformResult.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+case class ChangedFieldsUpdateTransformResult[C](updatedEntity: C,
+    fields: Seq[ChangedField[Any]]) {
+
+  /**
+   * Generates list of fields, for which two values do not match - i.e have been changed
+   * @return A list of fields for which values do not match
+   */
+  def getUpdatedFields(): Seq[String] = {
+    fields.filter({ case field => field.value1 != field.value2 }).map(_.fieldName).toSeq
```
Comment (GeorgiChochov) [inline menas/src/main/scala/za/co/absa/enceladus/rest/models/ChangedFieldsUpdateTransformResult.scala]:
```suggestion
    fields.filter(field => field.value1 != field.value2).map(_.fieldName)
```
No point in calling `.toSeq`, it already is a `Seq`.
Pick parens `()` or braces `{}`, not both.
No need for `case`.
Let's not create extra noise.

File: menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala (hunk 3 above)
Comment (GeorgiChochov) [reply menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala]: You can then use it in all create/update calls (as well as `addConformanceRule`, `updateDefaults`, `addDefault`, etc.)

File: menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala [hunk 5]
```diff
@@ -51,12 +63,21 @@ class DatasetService @Autowired()(datasetMongoRepository: DatasetMongoRepository
...
   }
 
-  def addConformanceRule(username: String, datasetName: String, rule: ConformanceRule): Future[Option[Dataset]] = {
-    super.update(username, datasetName) { dataset =>
-      dataset.copy(conformance = dataset.conformance :+ rule)
+  def addConformanceRule(username: String, datasetName: String, datasetVersion: Int, rule: ConformanceRule): Future[Option[Dataset]] = {
+    super.update(username, datasetName, datasetVersion, s"Conformance rule (${rule.order}) '${rule.outputColumn}' added.") { dataset =>
+      val updated = dataset.copy(conformance = dataset.conformance :+ rule)
+      
+      ChangedFieldsUpdateTransformResult(updatedEntity = updated, fields = Seq())
```
Comment (GeorgiChochov) [inline menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala]: Shouldn't this audit the change to the `conformance` field?

File: menas/src/main/scala/za/co/absa/enceladus/rest/services/AuditTrailService.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.mongodb.scala.Completed
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import za.co.absa.enceladus.rest.repositories._
+
+import scala.concurrent.Future
+
+@Service
+class AuditTrailService @Autowired()(auditTrailMongoRepository: AuditTrailMongoRepository)
```
Comment (GeorgiChochov) [inline menas/src/main/scala/za/co/absa/enceladus/rest/services/AuditTrailService.scala]: I would drop this class and have the `auditTrailRepository` where needed instead.

**Comment by Zejnilovic** [comment]
Unable to edit mapping tables
When I create a mapping table, I am not able to edit it. The edit doesn't increment the versions.
Stack:
```bash
org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:986)
	org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:881)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:855)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:116)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:84)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357)
	org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270)
</pre><p><b>Root Cause</b></p><pre>za.co.absa.enceladus.rest.exceptions.EntityAlreadyExistsException: Entity AAA (version. 0) already exists.
	za.co.absa.enceladus.rest.repositories.VersionedMongoRepository$$anonfun$update$1.apply(VersionedMongoRepository.scala:83)
	za.co.absa.enceladus.rest.repositories.VersionedMongoRepository$$anonfun$update$1.apply(VersionedMongoRepository.scala:81)
	scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
	scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	scala.concurrent.impl.CallbackRunnable.run(Promise.scala:36)
	scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by Zejnilovic** [comment]
Unable to create any conformance rule
Stack: 
```bash
org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:986)
	org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:881)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:855)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:116)
	org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:84)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357)
	org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270)
</pre><p><b>Root Cause</b></p><pre>za.co.absa.enceladus.rest.exceptions.EntityAlreadyExistsException: Entity Alfa (version. 10) already exists.
	za.co.absa.enceladus.rest.repositories.VersionedMongoRepository$$anonfun$update$1.apply(VersionedMongoRepository.scala:83)
	za.co.absa.enceladus.rest.repositories.VersionedMongoRepository$$anonfun$update$1.apply(VersionedMongoRepository.scala:81)
	scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
	scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	scala.concurrent.impl.CallbackRunnable.run(Promise.scala:36)
	scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by Zejnilovic** [comment]
Mapping table has `Schema` in audit trail table headers

image

**Comment by Zejnilovic** [comment]
Audit trails table on hover has background colour inconsistency

image

**Comment by Zejnilovic** [comment]
Changes is not scrollable, so if the change is bigger, I am not able to see full extent of it. Add line brakes or horizontal scroll bar

image

**Comment by Zejnilovic** [comment]
Audit trails table should be paginated. @lokm01 do we do it hear or add a feature request?

**Comment by Zejnilovic** [comment]
Audit trails table should be sortable by version or other. Both ASC and DESC. @lokm01 do we do it here or add a feature request

**Comment by Zejnilovic** [comment]
Audit trails URL endpoint is `menas/#/dataset/NAME/VERSION1`, which makes me troubled from the usability point of view. Let's say I am somewhere in the middle of scrolling and I misclick on a version and I am immediately transported to that version `VERSION2`. But when I press back, I get sent back to the `Basic Info` of the `VERSION1`, not the audit trails where I left off.

I think audit trails should be moved a bit from the general view of the dataset and have an endpoint like `menas/#/dataset/NAME/audit_trail` which is not dependant on versions.

**Comment by Zejnilovic** [comment]
Showing Dataset name in audit trails is in my opiniont waste of space. The name cannot be changed, must be unique and is right there in the header.

**Comment by lokm01** [comment]
> Audit trails table should be paginated. @lokm01 do we do it hear or add a feature request?

Let's do this as another issue.

**Comment by lokm01** [comment]
@Zejnilovic thanks! all of the above should be fixed now

**Comment by Zejnilovic** [comment]
> Showing Dataset name in audit trails is in my opiniont waste of space. The name cannot be changed, must be unique and is right there in the header.

I see this was done for mapping tables. How about for schema and dataset as well?

**Comment by Zejnilovic** [comment]
> Changes is not scrollable, so if the change is bigger, I am not able to see full extent of it. Add line brakes or horizontal scroll bar
> 
> image

`No Data` fields are still affected by this
image

**Comment by Zejnilovic** [comment]
> Changes is not scrollable, so if the change is bigger, I am not able to see full extent of it. Add line brakes or horizontal scroll bar
> 
> image

If the change is the creation of the field and the field has a long name, it still overflows.
image

**Comment by lokm01** [comment]
> > Showing Dataset name in audit trails is in my opiniont waste of space. The name cannot be changed, must be unique and is right there in the header.
> 
> I see this was done for mapping tables. How about for schema and dataset as well?

Good catch. Should be fixed now. 

Looking at the other feedback.

**Comment by lokm01** [comment]
> Audit trails table should be sortable by version or other. Both ASC and DESC. @lokm01 do we do it here or add a feature request

Lets do this as another feature. Ideally after pagination has been implemented as this will require server-side sorting.

**Comment by lokm01** [comment]
> Audit trails URL endpoint is `menas/#/dataset/NAME/VERSION1`, which makes me troubled from the usability point of view. Let's say I am somewhere in the middle of scrolling and I misclick on a version and I am immediately transported to that version `VERSION2`. But when I press back, I get sent back to the `Basic Info` of the `VERSION1`, not the audit trails where I left off.
> 
> I think audit trails should be moved a bit from the general view of the dataset and have an endpoint like `menas/#/dataset/NAME/audit_trail` which is not dependant on versions.
> 
> This goes hand to hand with the fact, that if I am on a `VERSION1` and I try to choose `VERSION1` again in audit trails, nothing happens. I expect to be moved to `Basic Info` or somehow "told" that the version is already selected.

Another issue please. This will require changing deep-link patterns etc.

**Comment by lokm01** [comment]
> > Audit trails URL endpoint is `menas/#/dataset/NAME/VERSION1`, which makes me troubled from the usability point of view. Let's say I am somewhere in the middle of scrolling and I misclick on a version and I am immediately transported to that version `VERSION2`. But when I press back, I get sent back to the `Basic Info` of the `VERSION1`, not the audit trails where I left off.
> > I think audit trails should be moved a bit from the general view of the dataset and have an endpoint like `menas/#/dataset/NAME/audit_trail` which is not dependant on versions.
> > This goes hand to hand with the fact, that if I am on a `VERSION1` and I try to choose `VERSION1` again in audit trails, nothing happens. I expect to be moved to `Basic Info` or somehow "told" that the version is already selected.
> 
> Another issue please. This will require changing deep-link patterns etc.

Created issue #239 for this.

File: conformance/pom.xml [hunk 7]
```diff
@@ -131,30 +131,21 @@
             </plugin>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-enforcer-plugin</artifactId>
-                <version>${maven.enforcer.version}</version>
+                <artifactId>maven-antrun-plugin</artifactId>
```
Comment (GeorgiChochov) [inline conformance/pom.xml]: Let's do this in a separate pull request. This is not part of the current task.

Comment (lokm01) [reply conformance/pom.xml]: Cool, done in PR #258

**Comment by lokm01** [comment]
> All my previous concerns fixed. I have new ones though.I am going to give them numbers so we can easily reference them
> 
> ## 1. Pretty output of schemas
> When we take in Schema, we take a JSON and turn it into a schema. Could we on the output prettify it? The audit trail is done for the user, what of it if the user cannot read it or needs to know scala/spark.
> 
> image
> 
> ## 2. Conformance rules shown as methods - understadnig them
> When we check on audit trails of conformance rules, they are printed as methods. Which is all nice, but the parameters arn't named, so how can the user know which field is which. Either name the parameters or create some doc, maybe as a static page to Menas?
> 
> image
> 
> ## 3. Conformance rules shown as methods - formating
> Changes are shown as methods. Which I really like, but could we add a space, a litteral space, betwen parameters of those methods. For readability purposes.
> 
> image

Opened issue #263 for these.

File: menas/ui/service/SchemaService.js [hunk 8]
```diff
@@ -15,66 +15,77 @@
...
-
-  this.updateSchema = function(sId, iVersion, sDesc) {
-    Functions.ajax("api/schema/edit", "POST", {
-      name: sId,
-      version: iVersion,
-      description: sDesc
-    }, function(oData) {
-      model.setProperty("/currentSchema", oData)
-      SchemaService.getSchemaList();
+
```
Comment (GeorgiChochov) [inline menas/ui/service/SchemaService.js]: What's up with the indentation changes? We settled on 2-spaces in `.js`. 
Maybe check for a `.editorconfig` plugin for eclipse?

File: menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala [hunk 9]
```diff
@@ -23,16 +23,18 @@ import za.co.absa.enceladus.rest.repositories.VersionedMongoRepository
...
 
-abstract class VersionedModelServiceTest[C <: VersionedModel] extends BaseServiceTest {
+abstract class VersionedModelServiceTest[C <: VersionedModel with Product with Auditable[C]] extends BaseServiceTest {
 
   val modelRepository: VersionedMongoRepository[C]
   val service: VersionedModelService[C]
 
   val millis100 = Duration(100, TimeUnit.MILLISECONDS)
-
+  val millis200 = Duration(100, TimeUnit.MILLISECONDS)
```
Comment (GeorgiChochov) [inline menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala]: This seems unnecessary.

File: menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala [hunk 10]
```diff
@@ -23,16 +23,18 @@ import za.co.absa.enceladus.rest.repositories.VersionedMongoRepository
...
+  
   private val validName = "validName"
-
+
```
Comment (GeorgiChochov) [inline menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala]: Whitespace added. I think your IDE isn't respecting the `.editoconfig` settings.

File: dataModel/src/main/scala/za/co/absa/enceladus/model/menas/audit/Auditable.scala [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+      fieldIndex.flatMap({ i =>
+          val newVal = newValue.productElement(i)
+          val oldVal = self.productElement(i)
+          if (newVal != oldVal)
+            Some(AuditTrailChange(field = name.declaredField, oldValue = Some(unwrapOption(oldVal).toString), newValue = Some(unwrapOption(newVal).toString), message = s"${name.humanReadableField} updated."))
+          else None
+      })
+
+    }).filter(_.isDefined).map(_.get)
```
Comment (GeorgiChochov) [inline dataModel/src/main/scala/za/co/absa/enceladus/model/menas/audit/Auditable.scala]:
```suggestion
    }).collect { 
      case Some(change) => change 
    }
```

File: menas/ui/service/SchemaService.js (hunk 8 above)
Comment (lokm01) [reply menas/ui/service/SchemaService.js]: Yeah doesn't support js, but this is me not doing autoformat, got it configured accordingly.

File: menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala (hunk 9 above)
Comment (lokm01) [reply menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala]: This is copy & paste error. Wanted 200ms as some of the tests failed previously on jenkins on future timeouts

File: dataModel/src/main/scala/za/co/absa/enceladus/model/menas/audit/Auditable.scala (hunk 11 above)
Comment (lokm01) [reply dataModel/src/main/scala/za/co/absa/enceladus/model/menas/audit/Auditable.scala]: yeah, better thanks :)

File: menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala (hunk 9 above)
Comment (GeorgiChochov) [reply menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala]: That 200ms still doesn't seem to be used anywhere though.

Comment (lokm01) [reply menas/src/test/scala/za/co/absa/enceladus/rest/services/VersionedModelServiceTest.scala]: It's used in the SchemaServiceTest - used it for now to test the audits and the parents.
